### PR TITLE
Kraus operator resolved with serial concatenation

### PR DESCRIPTION
**Pull request by Zshan0**
As requested in #3965

Made modifications to `kraus` in `kraus_protocol.py` by adding serial concatenation as mentioned.

Also added a test based on the issue.

Have Not implemented the precautions to avoid "large amount of combinatorial explosion" as I'm unclear on what "large amount" should be.

### CODE REVIEW

**Comment by tanujkhattar** [comment]
cc #4434 

@Zshan0 Can you please fix the failing tests?

File: cirq-core/cirq/protocols/kraus_protocol.py [hunk 1]
```diff
@@ -151,36 +155,55 @@ def kraus(
...
+
+        kraus_list = list(map(lambda x: _kraus_tensor(x, qubits, default), decomposed))
+        if not any([_check_equality(x, default) for x in kraus_list]):
+            kraus_result = kraus_list[0]
+            for i in range(1, len(kraus_list)):
+                kraus_result = [op_2.dot(op_1) for op_1 in kraus_result for op_2 in kraus_list[i]]
+                assert (
+                    len(kraus_result) < limit
+                ), f"{val} kraus decomposition had combinatorial explosion."
+            return tuple(kraus_result)
```
Comment (tanujkhattar) [inline cirq-core/cirq/protocols/kraus_protocol.py:198]: Use `apply_channel` and `apply_mixture` instead of rolling out your own serial concatenation logic.

**Review by tanujkhattar — CHANGES_REQUESTED** [review]
Left a few high level comments and gave more feedback offline. 

Tl;Dr - We should use the existing `apply_channel` and `apply_mixture` utility methods for concatenating channels and mixtures of decomposed operations -- similar to `apply_unitaries` method which uses `apply_unitary`.

File: cirq-core/cirq/protocols/kraus_protocol.py [hunk 2]
```diff
@@ -190,28 +207,71 @@ def has_kraus(val: Any, *, allow_decompose: bool = True) -> bool:
...
+                found = True
+                if val is None:
+                    val = kraus_list
+                else:
+                    val = tuple([np.kron(x, y) for x in val for y in kraus_list])
+
+        elif val is None:
+            val = (np.identity(2),)
+        else:
+            val = tuple([np.kron(x, np.identity(2)) for x in val])
```
Comment (tanujkhattar) [inline cirq-core/cirq/protocols/kraus_protocol.py:258]: I think this would not work for cases where qubits in `op_q` do not occur consecutively in the `qubits` array. For example, if qubits array is [1, 2, 3] and op is a CNOT on (1, 3) ? IIUC, it would behave as if op is a CNOT on (1, 2) instead ?

**Review by tanujkhattar — CHANGES_REQUESTED** [review]
A few high level comments which apply for changes in both kraus and mixture.
